tests/vertexonly/test_swarm.py

```diff
@@ -0,0 +1,136 @@
+    MPI.COMM_WORLD.Allgatherv(local_midpoints_size, local_midpoints_sizes)
+    midpoints = np.empty((num_cells, m.ufl_cell().geometric_dimension()), dtype=float)
+    MPI.COMM_WORLD.Allgatherv(local_midpoints, (midpoints, local_midpoints_sizes))
+    assert len(np.unique(midpoints, axis=0)) == len(midpoints)
+    return midpoints, local_midpoints
+
+
+@pytest.fixture(params=[pytest.param("interval", marks=pytest.mark.xfail(reason="swarm not implemented in 1d")),
+                        "square",
+                        pytest.param("extruded", marks=pytest.mark.xfail(reason="extruded meshes not supported")),
+                        "cube",
+                        "tetrahedron",
+                        pytest.param("immersedsphere", marks=pytest.mark.skip(reason="immersed parent meshes not supported and will segfault PETSc when creating the DMSwarm")),
+                        pytest.param("periodicrectangle", marks=pytest.mark.skip(reason="periodic meshes do not work properly with swarm creation"))])
+def parentmesh(request):
+    if request.param == "interval":
+        return UnitIntervalMesh(1)
+    elif request.param == "square":
+        return UnitSquareMesh(1, 1)
+    elif request.param == "extruded":
+        return ExtrudedMesh(UnitSquareMesh(1, 1), 1)
+    elif request.param == "cube":
+        return UnitCubeMesh(1, 1, 1)
+    elif request.param == "tetrahedron":
+        return UnitTetrahedronMesh()
+    elif request.param == "immersedsphere":
+        return UnitIcosahedralSphereMesh()
+    elif request.param == "periodicrectangle":
+        return PeriodicRectangleMesh(3, 3, 1, 1)
+
+
+# pic swarm tests
+
+def test_pic_swarm_in_plex(parentmesh):
+    """Generate points in cell midpoints of mesh `parentmesh` and check correct
+    swarm is created in plex."""
+
+    # Setup
+
+    parentmesh.init()
+    inputpointcoords, inputlocalpointcoords = cell_midpoints(parentmesh)
+    plex = parentmesh.topology._plex
+    from firedrake.petsc import PETSc
+    fields = [("fieldA", 1, PETSc.IntType), ("fieldB", 2, PETSc.ScalarType)]
+    swarm = mesh._pic_swarm_in_plex(plex, inputpointcoords, fields=fields)
+    # Get point coords on current MPI rank
+    localpointcoords = np.copy(swarm.getField("DMSwarmPIC_coor"))
+    swarm.restoreField("DMSwarmPIC_coor")
+    if len(inputpointcoords.shape) > 1:
+        localpointcoords = np.reshape(localpointcoords, (-1, inputpointcoords.shape[1]))
+    # Turn this into a number of points locally and MPI globally before
+    # doing any tests to avoid making tests hang should a failure occur
+    # on not all MPI ranks
+    nptslocal = len(localpointcoords)
+    nptsglobal = MPI.COMM_WORLD.allreduce(nptslocal, op=MPI.SUM)
+    # Get parent PETSc cell indices on current MPI rank
+    localparentcellindices = np.copy(swarm.getField("DMSwarm_cellid"))
+    swarm.restoreField("DMSwarm_cellid")
+
+    # Tests
+
+    # get custom fields on swarm - will fail if didn't get created
+    for field in fields:
```

Comment:
destructure `for name, size, dtype in fields`

Reply:
Done, much more readable. Thanks